### PR TITLE
Filter rLibDir by exists so that daemon.R references the correct file

### LINKED ISSUE #456: SparkR executor daemon fails to initialize when using conda
Tracking issue

On YARN, if the SparkR directory is not also added to SparkContext. On kubernetes with the base docker image.

### PR DESCRIPTION
**Pull request by jeremyjliu**
## Upstream SPARK-XXXXX ticket and PR link (if not applicable, explain)

Not filed in upstream, touches code for conda.

## What changes were proposed in this pull request?

rLibDir contains a sequence of possible paths for the SparkR package on the executor and is passed on to the R daemon with the SPARKR_RLIBDIR environment variable. This PR filters rLibDir for paths that exist before setting SPARKR_RLIBDIR, retaining existing functionality to preferentially choose a YARN or local SparkR install over conda if both are present.

See daemon.R: https://github.com/palantir/spark/blob/master/R/pkg/inst/worker/daemon.R#L23

Fixes #456 

## How was this patch tested?

Manually testing cherry picked on older version

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala [hunk 1]
```diff
@@ -360,10 +360,14 @@ private[r] object RRunner {
...
-    }.getOrElse(RUtils.sparkRPackagePath(isDriver = false))
-    val rExecScript = RUtils.sparkRInstallLocation(rLibDir, "/SparkR/worker/" + script)
+    val rLibDir = condaEnv.map(conda =>
+      RUtils.sparkRPackagePath(isDriver = false) :+ (conda.condaEnvDir + "/lib/R/library"))
+      .getOrElse(RUtils.sparkRPackagePath(isDriver = false))
+      .filter(dir => new File(dir).exists)
+    if (rLibDir.isEmpty) {
+      throw new SparkException("SparkR package is not installed on executor.")
+    }
+    val rExecScript = rLibDir.head + "/SparkR/worker/" + script
```
Comment (dvogelbacher) [inline core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: why this change? If both `RUtils.sparkRPackagePath(isDriver = false)` and `(conda.condaEnvDir + "/lib/R/library")` exist, we will just take the first one arbitrarily here and assume that it contains `"/SparkR/worker/" + script`.  That seems wrong?

Comment (jeremyjliu) [reply core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: Yeah, that was the existing logic that I'm preserving - if both sparkRPackagePath (either yarn or local rlib path) and conda package path exists, we defer to taking the former.

Comment (dvogelbacher) [reply core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: the existing code wouldn't have necessarily deferred to the former? Let's say:
- sparkRPackagePath = "dirA", this directory exists and is empty
- conda.condaEnvDir = "dirB", this directory exists and contains sub-directories `"/SparkR/worker/" + script`

Your code will now just use the sparkRPackagePath, trying to execute `dirA/SparkR/worker/" + script`, which will fail.
The old code would've checked which of the lib dirs actually contains `"/SparkR/worker/" + script`. So in this case it would actually choose to run `dirB/SparkR/worker/" + script` and succeed.

Comment (jeremyjliu) [reply core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: That's true. I'll adjust it so that it checks for the existence of script, like before. Though I will note that it's still possible for `worker.R` referenced in `daemon.R` to not exist in the rLibDir (and potentially other places in the code base). We will rely on the assumption that if the directory exists, then the proper sparkR files are present and throw loudly if this is not the case.

Comment (dvogelbacher) [reply core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: actually, if we really want to guarantee the same behaviour as before we need to filter for just dir exists in line 366 like you used to.
And then filter for `File(dir + "/SparkR/worker/" + script).exists` here and take head. Because even though those dirs might not contain the script they might contain libraries that we want to pass into the environment down below in line 379.

**Comment by jeremyjliu** [comment]
Closing in favor of https://github.com/palantir/spark/pull/460 so that circle tests can run.
